src/serac/numerics/functional/domain_integral_kernels.hpp

```diff
@@ -149,62 +154,120 @@ void evaluation_kernel_impl(trial_element_tuple trial_elements, test_element, do
+              (void)qf_state;
+              (void)qf_derivatives;
+              (void)update_state;
+              (void)qpts_per_elem;
+              (void)trial_elements;
+              (void)qf_inputs;
+              (void)interpolate_result;
+              (void)u;
+
+              [[maybe_unused]] static constexpr trial_element_tuple_type empty_trial_element{};
+              // batch-calculate values / derivatives of each trial space, at each quadrature point
+              (get<indices>(trial_elements)
+                   .interpolate(get<indices>(u)[elements[e]], rule, &get<indices>(interpolate_result[e]), ctx),
+               ...);
+              ctx.teamSync();
+
+              (promote_each_to_dual_when<indices == differentiation_index>(get<indices>(interpolate_result[e]),
+                                                                           &get<indices>(qf_inputs[e]), ctx),
+               ...);
+              ctx.teamSync();
+
+              // use J_e to transform values / derivatives on the parent element
+              // to the to the corresponding values / derivatives on the physical element
+              (parent_to_physical<get<indices>(empty_trial_element).family>(get<indices>(qf_inputs[e]), J, e, ctx),
+               ...);
+              ctx.teamSync();
+
+              // (batch) evalute the q-function at each quadrature point
+              //
+              // note: the weird immediately-invoked lambda expression is
+              // a workaround for a bug in GCC(<12.0) where it fails to
+              // decide which function overload to use, and crashes
+
+              auto qf_outputs = [&]() {
+                if constexpr (std::is_same_v<state_type, Nothing>) {
+                  return batch_apply_qf_no_qdata(qf, t, x[e], ctx, get<indices>(qf_inputs[e])...);
+                }
+                if constexpr (!std::is_same_v<state_type, Nothing>) {
+                  return batch_apply_qf(qf, t, x[e], &qf_state(e, 0), update_state, ctx, get<indices>(qf_inputs[e])...);
+                }
+              }();
+              ctx.teamSync();
+
+              // use J to transform sources / fluxes on the physical element
+              // back to the corresponding sources / fluxes on the parent element
+              physical_to_parent<test_element_type::family>(qf_outputs, J, e, ctx);
+
+              // write out the q-function derivatives after applying the
+              // physical_to_parent transformation, so that those transformations
+              // won't need to be applied in the action_of_gradient and element_gradient kernels
+              if constexpr (differentiation_index != serac::NO_DIFFERENTIATION) {
+                RAJA::RangeSegment x_range(0, leading_dimension(qf_outputs));
+                RAJA::loop<threads_x>(ctx, x_range, [&](int q) {
+                  qf_derivatives[e * uint32_t(qpts_per_elem) + uint32_t(q)] = get_gradient(qf_outputs[q]);
+                });
```

Comment:
[RAJA documentation](https://raja.readthedocs.io/en/develop/sphinx/user_guide/feature/loop_basic.html#complex-loops-raja-kernel) suggests against using nested RAJA::forall. Is RAJA::loop the same thing?